clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1220,10 +1220,14 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
      */
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
-        return partition != null ?
-                partition :
-                partitioner.partition(
-                        record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        if (partition == null) {
+            partition = partitioner.partition(record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        }
+        if (partition >= 0) {
+            return partition;
+        } else {
+            throw new IllegalStateException("Partition id is less than 0, partition_id=" + partition);
```

Comment:
I think it should be a KafkaException (or a subclass of it). IllegalStateException in this class refers to an invalid configuration that is more compile time than runtime, whereas a partitioning problem is purely runtime and only occurs with custom partitioners.

Reply:
I followed this error message type, to make it consistent https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java#L75